lib/puppet/environments.rb

```diff
@@ -213,9 +213,36 @@ def create_environment(name)
     end
 
     def valid_directory?(envdir)
-      name = Puppet::FileSystem.basename_string(envdir)
-      Puppet::FileSystem.directory?(envdir) &&
-         Puppet::Node::Environment.valid_name?(name)
+      return false unless Puppet::Node::Environment.valid_name?(Puppet::FileSystem.basename_string(envdir))
+
+      # Make an attempt to access the directory to determine if it's invalid
+      # due to permission problems or simply because it doesn't exist
+      begin
+        stat = Puppet::FileSystem.stat(envdir)
+      rescue Errno::ENOENT
+        false
```

Comment:
Should this be `return false` to short-circuit evaluation of the later `if` block?